src/transformers/models/t5/modeling_t5.py

```diff
@@ -951,9 +948,10 @@ def forward(
 
             # Model Parallel: If it's the last layer for that device, put things on the next device
             if self.model_parallel:
-                for k, v in self.device_map.items():
-                    if i == v[-1] and "cuda:" + str(k) != self.last_device:
-                        hidden_states = hidden_states.to("cuda:" + str(k + 1))
```

Comment:
and here we have an issue of an assumption that devices are not only ordered but also have a stride of 1, so the above would fail if we have `[0,2,1,3]`.

Reply:
the code is very hard to parse, that's why I think all these should be abstracted away, e.g. we should be able to say:
```
device_map.next_device(current_device)
```

Reply:
This work:
```
            def next_device(device_map, current_layer_number):
                devices = list(device_map.keys())
                for e, d in enumerate(devices):
                    if current_layer_number == self.device_map[d][-1]:
                        if e+1 != len(devices):
                            return devices[e+1]
                        else:
                            return devices[e]
                    
            # Model Parallel: If it's the last layer for that device, put things on the next device
            if self.model_parallel:
                hidden_states = hidden_states.to(next_device(self.device_map, i))
```

Reply:
This perhaps is more readable/efficient:
```
            def switch_to_next_device(device_map, current_layer_number):
                """ returns False if shouldn't switch, otherwise device id to switch to """
                devices = list(device_map.keys())
                for e, d in enumerate(devices):
                    if current_layer_number == device_map[d][-1] and e+1 != len(devices):
                        return devices[e+1] # next device
                return False
                    
            # Model Parallel: If it's the last layer for that device, put things on the next device
            if self.model_parallel:
                next_device_id = switch_to_next_device(self.device_map, i)
                if next_device_id is not False:
                    hidden_states = hidden_states.to(next_device_id)
```

Reply:
The other even simpler approach would be to create 2 reverse dicts for:

1. layer number -> device_id
2. layer number -> next_device_id 

then it's just a one trivial lookup at the point where it's needed, with the special case of the last group of layers pointing to None for next_device_id

Reply:
and if we stick to `cuda:X` we should do all that remapping at the beginning too and not have the mess of adding `cuda:` everywhere. And it's easier to compare foo.device too that way.

Reply:
Well, if #9323 is welcomed, this whole discussion while still valid becomes moot, since none of this code is needed at all.